tracker/src/index.rs

```diff
@@ -0,0 +1,331 @@
+    }
+
+    fn add_blocks(&mut self, old_height: u64, new_height: u64) {
+        for index in old_height + 1..new_height + 1 {
+            let hash = self.client.get_block_hash(index).unwrap();
+            self.add_next_block_to_index(hash).unwrap();
+        }
+    }
+
+    fn check_for_reorgs(&self) -> Option<ReorgInfo> {
+        let tip = self.checked_chain.last()?;
+        let last_tip_block = self.client.get_block_header_info(tip).unwrap();
+        if is_block_in_main_chain(&last_tip_block) {
+            return None;
+        }
+        let mut discarded_blocks = vec![tip.clone()];
+        for block_hash in self.checked_chain.iter().rev().skip(1) {
+            let block_header_info = self.client.get_block_header_info(&block_hash).unwrap();
+            if is_block_in_main_chain(&block_header_info) {
+                break;
+            } else {
+                discarded_blocks.push(block_header_info.hash);
+            }
+        }
+        let height_when_fork = self.height - discarded_blocks.len() as u64;
+        Some(ReorgInfo {
+            height_when_fork,
+            discarded_blocks,
+        })
+    }
+
+    fn add_next_block_to_index(&mut self, block_hash: BlockHash) -> Result<(), S::Err> {
+        let transactions = self.check_bitcoin_block_with_hash(block_hash.clone());
+        transactions
+            .into_iter()
+            .map(|tx| self.storage.store_record(tx))
+            .collect::<Result<Vec<()>, S::Err>>()?;
+        self.checked_chain.push(block_hash);
+        Ok(())
+    }
+
+    fn check_bitcoin_block_with_hash(&self, hash: BlockHash) -> Vec<Record> {
+        let block = self.client.get_block(&hash).unwrap();
+        let txs = block.txdata;
+
+        let mint_txs = txs
+            .into_iter()
+            .filter_map(|tx| self.parse_mint_transaction(hash, tx))
+            .collect();
+
+        mint_txs
+    }
+}
+
+impl<C: BitcoinClient, S: IndexStorage> Index<C, S> {
+    fn parse_mint_transaction(&self, block: BlockHash, tx: Transaction) -> Option<Record> {
+        let txid = tx.txid();
+        tx.output
+            .iter()
+            .enumerate()
+            .filter_map(|(out_pos, out)| {
+                self.parse_mint_output(out).map(|data| Record {
+                    bitcoin_block: block,
+                    bitcoin_tx_id: txid.clone(),
+                    bitcoin_output_position: out_pos as u64,
+                    data,
+                })
+            })
+            .next()
+    }
+
+    fn parse_mint_output(&self, out: &TxOut) -> Option<RecordData> {
+        let mut instructions = out.script_pubkey.instructions();
+
+        let first_instruction = instructions.next().and_then(|res| res.ok())?;
+        assert_instruction_return(first_instruction)?;
+
+        let push_bytes_instr = instructions.next().and_then(|res| res.ok())?;
+        let bag_id = parse_push_32_bytes(push_bytes_instr)?;
+
+        if !self.bags.iter().any(|bag| *bag == bag_id) {
```

Comment:
Is it possible to have bitcoin block scanned before the bags become known? Imagine we are bootstrapping the node and ingesting both chains. One network might be faster than the other.

Reply:
Resolved by https://github.com/runflame/mint/pull/1/commits/3f74bd4a8d30cbd8f788f02831f9599560ad8a89.